lib/CL/clEnqueueWriteBufferRect.c

```diff
@@ -90,37 +91,29 @@ POname(clEnqueueWriteBufferRect)(cl_command_queue command_queue,
     }
   assert(i < command_queue->context->num_devices);
 
-
-  /* execute directly */
-  /* TODO: enqueue the write_rect if this is a non-blocking read (see
-     clEnqueueWriteBuffer) */
-  if (command_queue->properties & CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE)
-    {
-      /* wait for the event in event_wait_list to finish */
-      POCL_ABORT_UNIMPLEMENTED("clEnqueueReadBufferRect: Out-of-order queue");
-    }
-  else
-    {
-      /* in-order queue - all previously enqueued commands must 
-       * finish before this read */
-      // ensure our buffer is not freed yet
-      POname(clRetainMemObject) (buffer);
-      POname(clFinish)(command_queue);
-    }
-
-  POCL_UPDATE_EVENT_RUNNING(event);
-
-  /* TODO: offset computation doesn't work in case the ptr is not 
-     a direct pointer */
-  device->ops->write_rect (device->data, ptr, 
-                           buffer->device_ptrs[device->dev_id].mem_ptr,
-                           buffer_origin, host_origin, region,
-                           buffer_row_pitch, buffer_slice_pitch,
-                           host_row_pitch, host_slice_pitch);
-
-  POCL_UPDATE_EVENT_COMPLETE(event);
-
-  POname(clReleaseMemObject) (buffer);
+  POname(clRetainMemObject) (buffer);
+
+  pocl_create_command (&cmd, command_queue, CL_COMMAND_WRITE_BUFFER_RECT,
+                       event, num_events_in_wait_list, event_wait_list, 1, 
+                       &buffer);
+
+  cmd->command.w_image.device_ptr = 
```

Comment:
Copying cmd->command.w_image in a shorter var would be easier to read and to modify if there is changes in future.